oximeter/collector/src/collection_task.rs

```diff
@@ -0,0 +1,894 @@
+    outbox: mpsc::Sender<CollectionTaskOutput>,
+
+    // Timer for making collections periodically.
+    collection_timer: Interval,
+
+    // Timer for reporting our own collection statistics to the database.
+    self_collection_timer: Interval,
+}
+
+impl CollectionTask {
+    // Construct a new collection task.
+    //
+    // This also spawns the internal task which itself manages the collections
+    // from our assigned producer. It then creates all the controlling queues
+    // for talking to this task and the inner task.
+    async fn new(
+        log: &Logger,
+        collector: self_stats::OximeterCollector,
+        producer: ProducerEndpoint,
+        outbox: mpsc::Sender<CollectionTaskOutput>,
+    ) -> (Self, mpsc::Sender<CollectionMessage>) {
+        // Create our own logger.
+        let log = log.new(o!(
+            "component" => "collection-task",
+            "producer_id" => producer.id.to_string(),
+        ));
+
+        // Setup queues for talking between ourselves, our controlling task
+        // handle, and the spawned collection loop itself.
+        let (task_tx, inbox) = mpsc::channel(N_QUEUED_TASK_MESSAGES);
+        let (producer_info_tx, producer_info_rx) = watch::channel(producer);
+        let (forced_collection_tx, forced_collection_rx) =
+            mpsc::channel(N_QUEUED_FORCED_COLLECTIONS);
+        let (timer_collection_tx, timer_collection_rx) =
+            mpsc::channel(N_QUEUED_TIMER_COLLECTIONS);
+        let (result_tx, result_rx) = mpsc::channel(N_QUEUED_RESULTS);
+        tokio::task::spawn(collection_loop(
+            log.clone(),
+            producer_info_rx,
+            forced_collection_rx,
+            timer_collection_rx,
+            result_tx,
+        ));
+
+        // Construct ourself, and return our controlling input queue.
+        let details = ProducerDetails::new(&producer);
+        let stats = self_stats::CollectionTaskStats::new(collector, &producer);
+        let collection_timer = Self::timer(producer.interval).await;
+        let self_collection_timer =
+            Self::timer(self_stats::COLLECTION_INTERVAL).await;
+        let self_ = Self {
+            log,
+            details,
+            stats,
+            inbox,
+            outbox,
+            producer_info_tx,
+            forced_collection_tx,
+            timer_collection_tx,
+            result_rx,
+            collection_timer,
+            self_collection_timer,
+        };
+        (self_, task_tx)
+    }
+
+    /// Helper to construct a timer and tick it.
+    ///
+    /// Since a `tokio::time::interval`'s first tick completes immediately, this
+    /// constructs the timer and then _ticks it_ once.
+    async fn timer(t: Duration) -> Interval {
+        let mut timer = interval(t);
+        timer.tick().await;
+        timer
+    }
+
+    /// Run the main loop of this collection task.
+    async fn run(mut self) -> TaskAction {
+        loop {
+            tokio::select! {
```

Comment:
Do we need to store the results of the `TaskAction` returned by whichever helper method we called and then act on it after the select? E.g., I think right now a `Shutdown` is ignored, because `handle_inbox_message` tells us `TaskAction::Break` and we don't look at it.

Reply:
I don't think `Shutdown` is ignored. We use `?` to propagate the returned value for `handle_inbox_message()`, so the `TaskAction::Break` does cause us to return. Does that seem right, or am I missing something?

Reply:
Hah! TIL `std::ops::ControlFlow` works with `?`. You're not missing anything; I skimmed right over the `?` without thinking about them.

Reply:
Yeah, it's great! I was going to return a `bool` or `Option<()>`, but the `ControlFlow` enum is very nice. A bit heavy-weight for this, but very convenient.